### PR TITLE
[edn/rtl] boot mode and auto req mode mutually exclusive

### PR DESCRIPTION
**Pull request by mwbranstad**
When initialized, only one hw mode can be active at a time.

Signed-off-by: Mark Branstad <mark.branstad@wdc.com>

### CODE REVIEW

**Review by senelson7 — APPROVED** [review]
Makes sense to me.

**Review by mcy — APPROVED** [review]
LG for software.

**Review by moidx — APPROVED** [review]

File: hw/ip/edn/rtl/edn_core.sv [hunk 1]
```diff
@@ -447,16 +447,16 @@ module edn_core import edn_pkg::*;
...
          (cmd_fifo_rst || !seq_auto_req_mode) ? '0 :
-         capt_gencmd_fifo_cnt ? (sfifo_gencmd_depth-1) :
-         capt_rescmd_fifo_cnt ? (sfifo_rescmd_depth-1) :
+         capt_gencmd_fifo_cnt ? (sfifo_gencmd_depth) :
+         capt_rescmd_fifo_cnt ? (sfifo_rescmd_depth) :
          (send_gencmd || send_rescmd)? (cmd_fifo_cnt_q-1) :
          cmd_fifo_cnt_q;
 
-  assign cmd_sent = (cmd_fifo_cnt_q == '0);
+  assign cmd_sent = (cmd_fifo_cnt_q == 13'h01);
```
Comment (vogelpi) [inline hw/ip/edn/rtl/edn_core.sv:455]: This change seems to be unrelated to boot and auto req modes. Is this a bug fix?

File: hw/ip/edn/rtl/edn_core.sv [hunk 2]
```diff
@@ -295,7 +295,7 @@ module edn_core import edn_pkg::*;
   // cmd req
   assign sw_cmd_req_load = reg2hw.sw_cmd_req.qe;
   assign sw_cmd_req_bus = reg2hw.sw_cmd_req.q;
-  assign auto_req_mode = reg2hw.ctrl.auto_req_mode.q;
+  assign auto_req_mode = (reg2hw.ctrl.hw_req_mode.q == 2'b01);
```
Comment (vogelpi) [inline hw/ip/edn/rtl/edn_core.sv:298]: Right now, the `hw_req_mode.q` value is tested just in two places and in a single file. So testing against the number literal sounds okay to me.

However, if you in the future plan to evaluate this signal in more places, or if you will introduce a separate mode signal that can indicate either of the two, I would recommend to define an enum type in `edn_pkg.sv`.

File: hw/ip/edn/data/edn.hjson [hunk 3]
```diff
@@ -84,24 +84,23 @@
...
-        { bits: "3",
-          name: "BOOT_REQ_DIS",
-          desc: '''
-                Setting this bit will disable the feature where the EDN block will automatically
+                2'b10 - BOOT_REQ_MODE:
+                This decode will enable the feature where the EDN block will automatically
                 send a boot-time request to the CSRNG application interface.
                 The purpose of this feature is to request entropy as fast as possible after reset,
                 and during chip boot-time.
+                2'b11 - Reserved.
```
Comment (vogelpi) [inline hw/ip/edn/data/edn.hjson:103]: Looking at the RTL, `2'b11` right now seems to behave just like `2'b00`. Do you want to state this in the documentation? Alternatively, but this will be more work, you could make this register or field `hwext` and then resolve reserved values to `2'b00`. This is what we do for AES.

File: hw/ip/edn/data/edn.hjson [hunk 4]
```diff
@@ -84,24 +84,23 @@
                 software before any further commands can be issued to these FIFOs.
                 '''
         }
-        { bits: "2",
-          name: "AUTO_REQ_MODE",
+        { bits: "3:2",
```
Comment (vogelpi) [inline hw/ip/edn/data/edn.hjson:87]: I recommend to do a proper enum here:
```
{ bits: "3:2",
  name: "HW_REQ_MODE",
  resval: ?
  desc: General description of the field, here I would list what happens if an invalid value is selected.
  enum: [
    { value: 0,
      name: "DISABLE"
      desc: mode specific description
    },
    { value: 1,
      name: "AUTO_REQ_MODE",
      desc: mode specific description
    },
    { value: "2",
      name: "BOOT_REQ_MODE",
      desc:
    },
    // No need to list the reserved value.
    // The regtool automatically prints a "Other values are reserved" statement.
  ]
```
This will look as follows in the online documentation: https://docs.opentitan.org/hw/ip/aes/doc/#Reg_ctrl_shadowed

**Review by vogelpi — APPROVED** [review]
This looks good to me @mwbranstad . I made some suggestions regarding the hjson (it supports an enum field).

File: hw/ip/edn/rtl/edn_core.sv (hunk 1 above)
Comment (mwbranstad) [reply hw/ip/edn/rtl/edn_core.sv:455]: Actually that is the point of this PR (a behavioral change) but is directly related to the number of concurrent operational modes allowed. This prompted the change register decode change.

File: hw/ip/edn/rtl/edn_core.sv (hunk 2 above)
Comment (mwbranstad) [reply hw/ip/edn/rtl/edn_core.sv:298]: I actually did what you are suggesting in csrng for the command command decode - multiple values used in many files. A very good choice it turned out. Here the decode is simple and isolated, think I will leave as is.

File: hw/ip/edn/data/edn.hjson (hunk 3 above)
Comment (mwbranstad) [reply hw/ip/edn/data/edn.hjson:103]: In most documentation that I have seen, a designation of reserved is not specified, mostly for future use. If the reserved field is programmed, it is assumed that unpredictable results could occur. Unless there is some security reason that I am not familiar with for AES, mapping the reserved values seems unneeded.

File: hw/ip/edn/data/edn.hjson (hunk 4 above)
Comment (mwbranstad) [reply hw/ip/edn/data/edn.hjson:87]: This is a much better format, I will change to this. Thanks @vogelpi for you review feedback!

**Comment by mwbranstad** [comment]
@tjaychen again @vogelpi added feedback that was incorporated. I think we can merge now.

File: hw/ip/edn/rtl/edn_core.sv (hunk 1 above)
Comment (vogelpi) [reply hw/ip/edn/rtl/edn_core.sv:455]: Ah I see. Thanks for the clarification.

File: hw/ip/edn/rtl/edn_core.sv (hunk 2 above)
Comment (vogelpi) [reply hw/ip/edn/rtl/edn_core.sv:298]: Sounds good to me!

File: hw/ip/edn/data/edn.hjson (hunk 3 above)
Comment (vogelpi) [reply hw/ip/edn/data/edn.hjson]: I am just worried about the worst case scenario where an attacker would flip a single bit from `2'b01` to `2'b11` and then EDN would enter an undefined state where it delivers constant "entropy". But it looks like the implementation currently wouldn't return anything / not acking the requests and the consumer would just starve. This is the way to go.

File: hw/ip/edn/data/edn.hjson (hunk 4 above)
Comment (vogelpi) [reply hw/ip/edn/data/edn.hjson:87]: You're welcome @mwbranstad , thanks for making the propsed changes :+1:
